blockchain/state.go

```diff
@@ -11,47 +10,77 @@ import (
-type ActiveState struct {
+// BeaconState is the state of a beacon block
+type BeaconState struct {
+	// Slot of last validator set change
+	ValidatorSetChangeSlot uint64
+	// List of validators
+	Validators []primitives.Validator
+	// Most recent crosslink for each shard
+	Crosslinks []primitives.Crosslink
+	// Last cycle-boundary state recalculation
+	// MEMO: old name in our code is LastStateRecalculationSlot
+	LastStateRecalculationSlot uint64
+	// Last finalized slot
+	LastFinalizedSlot uint64
+	// Last justified slot
+	LastJustifiedSlot uint64
+	// Number of consecutive justified slots
+	JustifiedStreak uint64
+	// Committee members and their assigned shard, per slot
+	ShardAndCommitteeForSlots [][]primitives.ShardAndCommittee
+	// Persistent shard committees
+	PersistentCommittees             []uint32
+	PersistentCommitteeReassignments []ShardReassignmentRecord
+	// Randao seed used for next shuffling
+	NextShufflingSeed chainhash.Hash
+	// Total deposits penalized in the given withdrawal period
+	DepositsPenalizedInPeriod []uint64
+	// Hash chain of validator set changes (for light clients to easily track deltas)
+	ValidatorSetDeltaHashChange chainhash.Hash
+	// Current sequence number for withdrawals
+	CurrentExitSeq uint64
+	// Genesis time
+	GenesisTime uint64
+	// PoW receipt root
+	ProcessedPowReceiptRoot  chainhash.Hash
+	CandidatePowReceiptRoots []CandidatePoWReceiptRootRecord
+	// Parameters relevant to hard forks / versioning.
+	// Should be updated only by hard forks.
+	PreForkVersion  uint64
+	PostForkVersion uint64
+	ForkSlotNumber  uint64
+	// Attestations not yet processed
 	PendingAttestations []transaction.Attestation
-	PendingActions      []transaction.Transaction
 	RecentBlockHashes   []chainhash.Hash
 	RandaoMix           chainhash.Hash
 }
 
-// State is active and crystallized state.
-type State struct {
-	Active       ActiveState
-	Crystallized CrystallizedState
+// ShardReassignmentRecord is the record of shard reassignment
+type ShardReassignmentRecord struct {
+	// Which validator to reassign
+	ValidatorIndex uint32
+	// To which shard
+	Shard uint64
+	// When
+	Slot uint64
 }
 
-// CrystallizedState is state that is updated every epoch
-type CrystallizedState struct {
-	ValidatorSetChangeSlot uint64
-	Crosslinks             []primitives.Crosslink
-	Validators             []primitives.Validator
-	LastStateRecalculation uint64
-	JustifiedStreak        uint64
-	LastJustifiedSlot      uint64
-	LastFinalizedSlot      uint64
-
-	// ShardAndCommitteeForSlots is an array of slots where each element
-	// is the committee assigned to that slot for each shard.
-	ShardAndCommitteeForSlots   [][]primitives.ShardAndCommittee
-	DepositsPenalizedInPeriod   []uint64
-	ValidatorSetDeltaHashChange chainhash.Hash
-	PreForkVersion              uint32
-	PostForkVersion             uint32
-	ForkSlotNumber              uint64
+// CandidatePoWReceiptRootRecord is the record of candidate PoW receipt root
```

Comment:
We don't have a PoW chain we'll being tying this to, so remove this and processedPoWReceiptRoot